pyomo/solvers/plugins/solvers/gurobi_direct.py

```diff
@@ -163,43 +163,31 @@ def _apply_solver(self):
     def _get_expr_from_pyomo_repn(self, repn, max_degree=2):
         referenced_vars = ComponentSet()
 
-        degree = canonical_degree(repn)
+        degree = repn.polynomial_degree()
         if (degree is None) or (degree > max_degree):
             raise DegreeError('GurobiDirect does not support expressions of degree {0}.'.format(degree))
 
-        if isinstance(repn, LinearCanonicalRepn):
-            if (repn.linear is not None) and (len(repn.linear) > 0):
-                list(map(referenced_vars.add, repn.variables))
-                new_expr = self._gurobipy.LinExpr(repn.linear, [self._pyomo_var_to_solver_var_map[i] for i in repn.variables])
-            else:
-                new_expr = 0
+        if len(repn.linear_vars) > 0:
+            referenced_vars.update(repn.linear_vars)
+            new_expr = self._gurobipy.LinExpr(repn.linear_coefs, [self._pyomo_var_to_solver_var_map[i] for i in repn.linear_vars])
+        else:
+            new_expr = 0.0
 
-            if repn.constant is not None:
-                new_expr += repn.constant
+        for i,v in enumerate(repn.quadratic_vars):
+            x,y = v
+            new_expr += repn.quadratic_coefs[i] * self._pyomo_var_to_solver_var_map[x] * self._pyomo_var_to_solver_var_map[y]
+            referenced_vars.add(x)
+            referenced_vars.add(y)
 
-        else:
-            new_expr = 0
-            if 0 in repn:
-                new_expr += repn[0][None]
-
-            if 1 in repn:
-                for ndx, coeff in repn[1].items():
-                    new_expr += coeff * self._pyomo_var_to_solver_var_map[repn[-1][ndx]]
-                    referenced_vars.add(repn[-1][ndx])
-
-            if 2 in repn:
-                for key, coeff in repn[2].items():
-                    tmp_expr = coeff
-                    for ndx, power in key.items():
-                        referenced_vars.add(repn[-1][ndx])
-                        for i in range(power):
-                            tmp_expr *= self._pyomo_var_to_solver_var_map[repn[-1][ndx]]
-                    new_expr += tmp_expr
+        new_expr += repn.constant
 
         return new_expr, referenced_vars
 
     def _get_expr_from_pyomo_expr(self, expr, max_degree=2):
-        repn = generate_canonical_repn(expr)
+        if max_degree == 2:
+            repn = generate_standard_repn(expr, quadratic=True)
+        else:
+            repn = generate_standard_repn(expr, quadratic=False)
```

Comment:
The max_degree is determined by the version of Gurobi. Older versions of Gurobi do not support quadratic constraints. If an older version of Gurobi is being used and max_degree is 1, will "quadratic=False" cause an error to be raised if a user tries to use quadratic constraints, or will this quietly ignore the quadratic parts of the expression?

Reply:
Bill and I discussed this in person. This is handled correctly.